### PR TITLE
test: fix CONTRIBUTING test instructions

### PR DESCRIPTION
**Pull request by aakropotkin**
## Proposed Changes

Various improvements to the test suit and harnesses.
Aligns advice in `CONTRIBUTING.md` with latest build flows.
Makes it possible to use builds of `flox` in local dirs in test suite.

## Current Behavior

Advice in `CONTRIBUTING.md` does not work.
It's not possible to pass flags to `bats`.
It's not possible to run tests on local builds like `./target/debug/flox`.

## Checks

- [x] All tests pass.
- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] I have accepted the flox [Contributor License Agreement](../blob/main/.github/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](../blob/main/.github/CONTRIBUTORS.csv) file by way of this pull request or one done previously.

### CODE REVIEW

**Review by garbas — APPROVED** [review]
👍 I think we should allocate some time to rework some of our tests since they leak (looking at you `integration.bats 😄) and also hopefully we will be able to run the tests in parallel. But that is for the future.

File: pkgs/flox-tests/default.nix [hunk 1]
```diff
@@ -54,10 +54,12 @@ in
 
         if command -v flox &> /dev/null
         then
-          export TMP_FLOX_CLI=$(command -v flox)
+          TMP_FLOX_CLI="$(command -v flox)"
```
Comment (garbas) [inline pkgs/flox-tests/default.nix:57]: @aakropotkin Why is this better than having it in one line?

File: pkgs/flox-tests/default.nix [hunk 2]
```diff
@@ -79,78 +81,47 @@ in
...
-        for arg in "$@"; do
-          shift
-          case "$arg" in
-            '--flox')   set -- "$@" '-F';;
-            '--tests')  set -- "$@" '-T';;
-            '--watch')  set -- "$@" '-W';;
-            '--help')   set -- "$@" '-h';;
-            *)          set -- "$@" "$arg" ;;
+        WATCH=
+        while [[ "$#" -gt 0 ]]; do
```
Comment (garbas) [inline pkgs/flox-tests/default.nix:85]: Magic 🪄

File: pkgs/flox-tests/default.nix [hunk 3]
```diff
@@ -79,78 +81,47 @@ in
...
+        : "''${FLOX_CLI:=$TMP_FLOX_CLI}";
+        : "''${TESTS_DIR:=$PWD/tests}";
+        export TESTS_DIR FLOX_CLI;
 
         # isolate git config
-        ssh-keygen -t ed25519 -q -N "" -f $TEMP_FLOX/id_ed25519
-        git config --global gpg.format ssh
-        git config --global user.signingkey $TEMP_FLOX/id_ed25519.pub
+        ssh-keygen -t ed25519 -q -N "" -f "$TEMP_FLOX/id_ed25519";
+        # FIXME: Why on earth is the test suite modifying my GPG perms?!
```
Comment (garbas) [inline pkgs/flox-tests/default.nix:104]: Yeah, this should be moved into the setup part of the test suit (without the `--global` or isolate `git` better)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -169,34 +159,50 @@ squashed into multiple distinct change sets.
...
 Integration tests are written with `bats` and `expect`.
-They are located in the `./tests` folder.
+They are located in the `<flox>/tests` folder.
 To run them:
 
 ```console
-$ cargo build
-$ flox run .#flox-tests --flox ./target/debug/flox
+$ flox develop flox --command 'cargo build';
+$ flox shell '.#flox-tests' --command flox-tests --flox ./target/debug/flox;
```
Comment (ysndr) [inline CONTRIBUTING.md:178]: what's wrong with run here?
I dont think we even documented flox shell fwiw having that as part of our own testing seems counterintuitive

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -169,34 +159,50 @@ squashed into multiple distinct change sets.
...
+You can pass arbitrary flags through to `bats` using a `--` separator - however
+bugs in the `flox` CLI parser require you to use `sh -c` to wrap the command.
+Failing to wrap will cause `flox` to "consume" the `--` rather than pass it
+through to the inner command:
 
 ```console
-$ flox run .#flox-tests --flox ./target/debug/flox --watch -- --filter "flox search"
+$ flox develop flox --command 'cargo build';
+$ flox shell '.#flox-tests' --command                                   \
+    sh -c 'flox-tests --flox ./target/debug/flox -- ./tests/run.bats';
```
Comment (ysndr) [inline CONTRIBUTING.md:201]: This looks even more like it shouldn't be necessary

File: pkgs/flox-tests/default.nix [hunk 6]
```diff
@@ -79,78 +81,47 @@ in
...
+        #        This took me so long to debug...
+        #        I'm leaving it for now but this is high on the TODO list.
+        git config --global gpg.format ssh;
+        git config --global user.signingkey "$TEMP_FLOX/id_ed25519.pub";
```
Comment (ysndr) [inline pkgs/flox-tests/default.nix:108]:
```suggestion
        export GIT_CONFIG_GLOBAL="$TEMP_FLOX/gitconfig";
        git config --global user.name "Flox Integration;
        git config --global user.email "integration@localhost;
        git config --global gpg.format ssh;
        git config --global user.signingkey "$TEMP_FLOX/id_ed25519.pub";
```

**Review by ysndr — COMMENTED** [review]
This seems to do way more than fixing the contribution instructions and could possibly be split up a bit but if you dont split it up lets fix the gpg thing in here too, or it gets forgotten real quick

File: pkgs/flox-tests/default.nix (hunk 1 above)
Comment (aakropotkin) [reply pkgs/flox-tests/default.nix:57]: `export` with `"$( ... )"` has undefined behavior when used in combination with `set -e` and/or `set -o pipefail`.

Some shells will fail to export the variable as the empty string without exiting failure. By splitting the `export` to a separate line error catching will work properly.

This is a bit similar to how `: CMD` works.

File: CONTRIBUTING.md (hunk 4 above)
Comment (aakropotkin) [reply CONTRIBUTING.md:178]: `flox run` doesn't work here. I tried it several ways and the `flox run` arg parser will crash. Particularly try running with `flox run CMD -- ARGS`.

File: CONTRIBUTING.md (hunk 5 above)
Comment (aakropotkin) [reply CONTRIBUTING.md:201]: I agree that I dislike how this had to be written. The arg parser for `flox run` needs fixes.

File: pkgs/flox-tests/default.nix (hunk 2 above)
Comment (aakropotkin) [reply pkgs/flox-tests/default.nix:85]: <3

File: pkgs/flox-tests/default.nix (hunk 6 above)
Comment (aakropotkin) [reply pkgs/flox-tests/default.nix]: I was wrong `gitconfig` was already handled here.
The issue is in the `gpgsign` test which I'll fix later.

**Comment by aakropotkin** [comment]
> +1 I think we should allocate some time to rework some of our tests since they leak (looking at you `integration.bats smile) and also hopefully we will be able to run the tests in parallel. But that is for the future.

The good news is that we can parallelize many of the tests I was cleaning up.
The use of `setup` and `teardown` for individual tests is incredibly useful but it only really works if you aren't depending on the state of past tests.

File: CONTRIBUTING.md (hunk 4 above)
Comment (ysndr) [reply CONTRIBUTING.md:178]: you nee two `--` one for flox and one for nix...

File: CONTRIBUTING.md (hunk 5 above)
Comment (ysndr) [reply CONTRIBUTING.md:201]: yep thats nix' parser crashing

File: CONTRIBUTING.md (hunk 4 above)
Comment (ysndr) [reply CONTRIBUTING.md:178]: `flox run .#flox-tests -- -- --flox ./target/debug/flox`

Comment (ysndr) [reply CONTRIBUTING.md:178]: the other one is fun `flox run .#flox-tests -- -- --flox ./target/debug/flox -- ./tests/run.bats`

Comment (ysndr) [reply CONTRIBUTING.md:178]: how to work with three successive arg parsers ^^

**Comment by aakropotkin** [comment]
> This seems to do way more than fixing the contribution instructions and could possibly be split up a bit but if you dont split it up lets fix the gpg thing in here too, or it gets forgotten real quick

Lol merged

**Pull request merged**